### PR TITLE
First steps of the Problem Handling refactor to the new design

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Implements:
- https://www.pivotaltracker.com/story/show/184226137
- https://www.pivotaltracker.com/story/show/184226434
- https://www.pivotaltracker.com/story/show/184226462

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: test/Table_Tests/src/Database/Common_Spec.enso [hunk 1]
```diff
@@ -84,13 +84,6 @@ spec prefix connection =
             r_1 = df.order_by (Sort_Column_Selector.By_Name [Sort_Column.Name "quantity", Sort_Column.Name "price" Sort_Direction.Descending])
             r_1.at 'id' . to_vector . should_equal [4,2,3,1,6,5]
 
-        Test.specify 'should return warnings and errors when passed a non-existent column' <|
-            action = df.order_by (Sort_Column_Selector.By_Name [Sort_Column.Name 'foobar']) on_problems=_
-            tester table =
-                table.at 'id' . to_vector . should_equal [1,2,3,4,5,6]
-            problems = [Missing_Input_Columns.Error [Sort_Column.Name 'foobar'], No_Input_Columns_Selected]
-            Problems.test_problem_handling action problems tester
```
Comment (radeusgd) [inline test/Table_Tests/src/Database/Common_Spec.enso:92]: This test has been superseded by tests in `Order_By_Spec`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso [hunk 2]
```diff
@@ -48,15 +49,24 @@ type Map key value
...
 
              example_from_vector = Map.from_vector [["A", 1], ["B", 2]]
-    from_vector : Vector Any -> Map
-    from_vector vec = vec.fold Map.empty (m -> el -> m.insert (el.at 0) (el.at 1))
+    from_vector : Vector Any -> Boolean -> Map
+    from_vector vec allow_duplicates=False =
+        vec.fold Map.empty m-> el->
+            key = el.at 0
+            value = el.at 1
+            if allow_duplicates.not && m.contains_key key then Error.throw (Illegal_Argument.Error "`Map.from_vector` encountered duplicate key: "+key.to_display_text) else                m.insert key value
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso]:
```suggestion
            if allow_duplicates.not && m.contains_key key then Error.throw (Illegal_Argument.Error "`Map.from_vector` encountered duplicate key: "+key.to_display_text) else
                m.insert key value
```

Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso]: could flip the if to make it clearer to read.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 3]
```diff
@@ -133,19 +133,16 @@ type Table
...
            `Missing_Input_Columns`.
-         - If duplicate columns, names or indices are provided, a
-           `Duplicate_Column_Selectors`.
          - If a column index is out of range, a `Column_Indexes_Out_Of_Range`.
-         - If two distinct indices would refer to the same column, a
-           `Input_Indices_Already_Matched`, indicating that the additional
-           indices will not introduce additional columns.
          - If there are no columns in the output table, a `No_Output_Columns` is
            raised as an error regardless of the problem behavior, because it is
            not possible to create a table without any columns.
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:148]: As these are now errors should we have them as `!` on the type signature.

**Review by jdunkerley — APPROVED** [review]

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso (hunk 3 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:148]: I'll amend this in the next pass along with the doc update.

**Pull request merged**